fieldservice_sale/models/sale_order.py

```diff
@@ -25,10 +25,76 @@ class SaleOrder(models.Model):
     @api.depends('order_line.product_id')
     def _compute_fsm_order_ids(self):
         for order in self:
-            order.fsm_order_ids = self.env['fsm.order'].search([
+            orders = self.env['fsm.order']
+            orders += self.env['fsm.order'].search([
                 ('sale_line_id', 'in', order.order_line.ids)])
+            orders += self.env['fsm.order'].search([
+                ('sale_id', '=', order.id)])
+            order.fsm_order_ids = orders
             order.fsm_order_count = len(order.fsm_order_ids)
 
+    def _field_create_fsm_order_prepare_values(self):
+        self.ensure_one()
+        return {
+            'customer_id': self.partner_id.id,
+            'location_id': self.fsm_location_id.id,
+            # 'request_early': ,
+            'scheduled_date_start': self.date_fsm_request,
+            'description': self.name,
+            # 'template_id': self.product_id.fsm_order_template_id.id,
+            'sale_id': self.id,
+            'company_id': self.company_id.id,
+        }
+
+    @api.multi
+    def _field_create_fsm_order(self):
+        """ Generate fsm_order for the given Sale Order, and link it.
+            :return a mapping with the sale order id and its linked fsm_order
+            :rtype dict
+        """
+        result = {}
+        for so in self:
+            # create fsm_order
+            values = so._field_create_fsm_order_prepare_values()
+            fsm_order = self.env['fsm.order'].sudo().create(values)
+            # post message on SO
+            msg_body = _(
+                """Field Service Order Created: <a href=
+                   # data-oe-model=fsm.order data-oe-id=%d>%s</a>
+                """) % (fsm_order.id, fsm_order.name)
+            so.message_post(body=msg_body)
+            # post message on fsm_order
+            fsm_order_msg = _(
+                """This order has been created from: <a href=
+                   # data-oe-model=sale.order data-oe-id=%d>%s</a>
+                """) % (so.id, so.name)
+            fsm_order.message_post(body=fsm_order_msg)
+            result[so.id] = fsm_order
+        return result
+
+    @api.multi
+    def _field_find_fsm_order(self):
+        """ Find the fsm_order generated by the Sale Order. If no fsm_order
```

Comment:
I think it's better to adapt the same behavior as delivery orders : when you cancel the sale, the picking is cancelled, when set to quotation then confirmed again a new picking is created.

Reply:
That is probably a better method than this is being done here.  I agree there is a fair amount of code duplication between sale_order and sale_order_line.  While similar they are also a little different.

I originally modeled this after v11 of sale_timesheet.